src/character.cpp

```diff
@@ -1936,6 +1936,7 @@ bool Character::pour_into( item &container, item &liquid )
         auto qty = std::min( liquid.charges, container.ammo_capacity() - container.ammo_remaining() );
         liquid.charges -= qty;
         container.ammo_set( liquid.typeId(), container.ammo_remaining() + qty );
+        container.on_contents_changed();
```

Comment:
Mostly OK.
It works for the common use cases and more generic, automatic access (visiting) is explicitly prevented.
At the moment it needs special casing for all relevant cases.

It has the tiny advantage of explicitly marking "here the code tampers with contents" sections.